### PR TITLE
Fix groupby binary ops when grouped array is subset relative to other

### LINKED ISSUE #7797: More `groupby` indexing problems
### What happened?

There is still something wrong with the groupby indexing changes from `2023.04.0` onward. Here are two typical ways of doing a seasonal anomaly calculation, which return the same result on `2023.03.0` but are different on `2023.04.2`:

```python
import numpy as np
import xarray as xr

# monthly timeseries that should return "zero anomalies" everywhere
time = xr.date_range("2023-01-01", "2023-12-31", freq="MS")
data = np.linspace(-1, 1, 12)
x = xr.DataArray(data, coords={"time": time})
clim = xr.DataArray(data, coords={"month": np.arange(1, 13, 1)})

# seems to give the correct result if we use the full x, but not with a slice
x_slice = x.sel(time=["2023-04-01"])

# two typical ways of computing anomalies
anom_gb = x_slice.groupby("time.month") - clim
anom_sel = x_slice - clim.sel(month=x_slice.time.dt.month)

# passes on 2023.3.0, fails on 2023.4.2
# the groupby version is aligning the indexes wrong, giving us something other than 0
assert anom_sel.equals(anom_gb)
```

Related: #7759 #7766
cc @dcherian 

### What did you expect to happen?

_No response_

### Minimal Complete Verifiable Example

_No response_

### MVCE confirmation

- [ ] Minimal example — the example is as focused as reasonably possible to demonstrate the underlying issue in xarray.
- [ ] Complete example — the example is self-contained, including all data and the text of any traceback.
- [ ] Verifiable example — the example copy & pastes into an IPython prompt or [Binder notebook](https://mybinder.org/v2/gh/pydata/xarray/main?urlpath=lab/tree/doc/examples/blank_template.ipynb), returning the result.
- [ ] New issue — a search of GitHub Issues suggests this is not a duplicate.

### Relevant log output

_No response_

### Anything else we need to know?

_No response_

### Environment

<details>

</details>

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Closes #7797
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

cc @slevang

### CODE REVIEW

File: xarray/tests/test_groupby.py [hunk 1]
```diff
@@ -2289,3 +2296,20 @@ def test_resample_cumsum(method: str, expected_array: list[float]) -> None:
     actual = getattr(ds.foo.resample(time="3M"), method)(dim="time")
     expected.coords["time"] = ds.time
     assert_identical(expected.drop_vars(["time"]).foo, actual)
+
+
+def test_groupby_binary_op_regression():
```
Comment (headtr1ck) [inline xarray/tests/test_groupby.py:2301]:
```suggestion
def test_groupby_binary_op_regression() -> No e:
```

Annotate new tests plz :)

**Review by headtr1ck — COMMENTED** [review]
I think the broken test was actually wrong before, now the actual result looks more like what I would expect.

**Comment by slevang** [comment]
Thanks for the quick fix! Not sure about the bitshift test but I'm assuming @headtr1ck is right.

**Comment by abrammer** [comment]
Apologies, that's my bad.  Looks like I introduced a broken test and didn't manually double check the results coming back.   The right shift test should have been:

```    python
    right_expected = Dataset(
        {
            "x": ("index", [0, 0, 2, 2]),
            "y": ("index", [-1, -1, -2, -2]),
            "level": ("index", [0, 0, 4, 4]),
            "index": [0, 1, 2, 3],
        }
    )

    right_actual = (left_expected.groupby("level") >> shift).reset_coords(names="level")
    assert_equal(right_expected, right_actual)
```

File: doc/whats-new.rst [hunk 2]
```diff
@@ -39,7 +39,8 @@ Deprecations
 
 Bug fixes
 ~~~~~~~~~
-
+- Fix groupby binary ops when grouped array is subset relative to other. (:issue:`7797`).
+By `Deepak Cherian <https://github.com/dcherian>`_.
```
Comment (dcherian) [inline doc/whats-new.rst:43]:
```suggestion
  By `Deepak Cherian <https://github.com/dcherian>`_.
```

**Comment by slevang** [comment]
Would it be possible to run another bug fix release with this incorporated? Or I guess we're already on to `2023.5.0` according to the date.

**Pull request merged**